File: R codes/SARSCOV_dist.R
# Load data
df <- read.csv("C:/Users/ADRIJA/OneDrive/Documents/SARS COV 2 VARIANTS.csv", stringsAsFactors = FALSE)

variants <- df$Variant
genomes <- df$Genome
n <- length(variants)
dist_matrix <- matrix(0, n, n, dimnames = list(variants, variants))

# Two-row Wagner–Fischer
wagner_fischer_two_row <- function(a, b) {
  if (nchar(a) < nchar(b)) {
    tmp <- a; a <- b; b <- tmp
  }
  m <- nchar(a); n <- nchar(b)
  prev <- 0:n
  for (i in seq_len(m)) {
    curr <- integer(n + 1)
    curr[1] <- i
    for (j in seq_len(n)) {
      cost <- ifelse(substr(a, i, i) == substr(b, j, j), 0, 1)
      curr[j + 1] <- min(
        prev[j + 1] + 1,
        curr[j] + 1,
        prev[j] + cost
      )
    }
    prev <- curr
  }
  return(prev[n + 1])
}

# Compute normalized distances
for (i in seq_len(n)) {
  for (j in seq_len(n)) {
    if (i < j) {
      d <- wagner_fischer_two_row(genomes[i], genomes[j])
      dist_matrix[i, j] <- d / (nchar(genomes[i]) + nchar(genomes[j]))
    }
  }
}
for(i in seq_len(n)){
  for(j in seq_len(n)){
    if(i>j){
      dist_matrix[i,j]<-dist_matrix[j,i]
    }
  }
}
write.csv(dist_matrix, "SARS_COV_2_Variant_Distance_Matrix_R.csv", row.names = TRUE)

File: Python Codes/SARSCOV_dist.py
import pandas as pd
import numpy as np

# ---------- Step 1: Read Data ----------
df = pd.read_csv("C:/Users/ADRIJA/OneDrive/Documents/SARS-COV 2 VARIANTS.csv")

# Assuming columns are named "Variant" and "Full_Genome"
variants = df["Variant"].tolist()
genomes = df["Genome"].tolist()

n = len(variants)
matrix = np.zeros((n, n))

# ---------- Step 2: Two-row Wagner–Fischer Algorithm ----------
def wagner_fischer_two_row(a: str, b: str) -> int:
    """Compute Levenshtein distance using O(min(len(a), len(b))) space."""
    if len(a) < len(b):
        a, b = b, a  # ensure a is the longer string
    prev = list(range(len(b) + 1))
    for i in range(1, len(a) + 1):
        curr = [i] + [0] * len(b)
        for j in range(1, len(b) + 1):
            cost = 0 if a[i - 1] == b[j - 1] else 1
            curr[j] = min(prev[j] + 1, curr[j - 1] + 1, prev[j - 1] + cost)
        prev = curr
    return prev[-1]

# ---------- Step 3: Compute Normalized Distance Matrix ----------
for i in range(n):
    for j in range(n):
        if i == j:
            matrix[i, j] = 0.0
        else if i<j:
            d = wagner_fischer_two_row(genomes[i], genomes[j])
            matrix[i, j] = d / (len(genomes[i]) + len(genomes[j]))
for i in range(n):
    for j in range(i):
        matrix[i,j]=matrix[j,i]
# ---------- Step 4: Create and Save Matrix ----------
distance_df = pd.DataFrame(matrix, index=variants, columns=variants)
distance_df.to_csv("Variant_Distance_Matrix.csv", index=True)

print(distance_df)
